keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +594,165 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
+    x:
+      x-kubernetes-validation:
+      - rule: "self.all(element, element.y.startsWith('prefix:'))"
+      type: array
+      items:
+        type: object
+        properties:
+          name: y
+          type: string
+          name: z
+          type: string
+```
+
+The worst case list size of `x` would be based on the size of an object with two strings (`y` and `z`).
+
+We will also account for where in a schema CEL expression is placed when calculating its cost. For example:
+
+```
+spec:
+  type: object
+  properties:
+    x:
+      type: array
+      items:
+        type: object
+        properties:
+          name: y
+          type: string
+          x-kubernetes-validation:
+          - rule: "self.startsWith('prefix:')"
+```
+
+In this example the validation rule is contained within a list, so its expression cost would be `cost(startsWith) * worst_case_length(x)`. Note that this means some of the cost calculation will
+happen outside of the CEL cost function (which would only calculate the `startsWith` cost).
+
+During CRD validation, when an expression exceeds the cost limit, the error message will include both the limit and the cost of the expression to facilitate debugging. The cost calculation rules will be documented.
+
+#### Runtime Cost Budget
+
+We will instrument CEL to increment a "cost" counter during CEL expression evaluation. The costs for 
+operations will be the same as used to compute Estimated Costs. The main difference is that the cost of list 
+iteration and branch operations (ternary, short circuited ORs) will be measured during evaluation and won't 
+need to be based on worst case estimates. If the cost exceeds a provided cost limit, CEL evaluation will be halted.
+
+We will integrate with the API Priority and Fairness system, providing with information it needs to be 
+informed about CEL resource utilization and using the information it provides back to set runtime cost limits.
+
+### Request lifetime Bound
+
+We will wire in the request context into CEL so that CEL will halt expression evaluation if the context is 
+canceled.
+
+### Bounds
+
+Based on our current data we believe the limits should be:
+
+- per-expression cost: 8,000,000
+- per-request cost: 800,000,000
+- per-expression budget: Selected by API Priority and Fairness
+- per-request timeout limit: tied to request lifetime via the request context
+
+The 8,000,000 figure is based around a worst case scenario of a list of 2,000,000 elements iterated through
+in O(n) fashion with a loop body with a cost of 4 (enough for a basic regex). Since basic CEL expressions have 
+a cost of 1, this should be plenty in most cases, but `maxLength` can be set on any lists where more
+computation needs to be done. This figure will also prevent things such as expressions from accidentally
+being O(n<sup>2</sup>) or worse; the cost for iterating in O(n<sup>2</sup>) fashion over a list of only
+30,000 elements with a loop body of cost 1 is about 36,000,000.
```

Comment:
seems like we can indicate this will be pinned down at impl time experimentally with benchmarking rather than committing to specific numbers here